dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/infer/DescriptorLanguageInferrer.java

```diff
@@ -0,0 +1,204 @@
+        // Without including language/file-type-specific logic such as parsing, which we are specifically trying to avoid in the
+        // .dockstore.yml inference code, we have no way of conclusively differentiating the two cases.
+        // So, we process each full chunk, then split the full chunk into subchunks at the spaces, and process each subchunk.
+        // That way, we handle both cases, and any spurious computed paths will very likely be ignored during subsequent processing,
+        // because they do not point at actual files.
+        for (String chunk: NON_SPACE_SEPARATORS.split(content)) {
+            if (isReferencedPath(chunk)) {
+                toAbsolutePath(srcPath, chunk).ifPresent(paths::add);
+            }
+            for (String subChunk: chunk.split(SPACE)) {
+                if (isReferencedPath(subChunk)) {
+                    toAbsolutePath(srcPath, subChunk).ifPresent(paths::add);
+                }
+            }
+        }
+        return paths;
+    }
+
+    protected boolean isReferencedPath(String possiblePath) {
+        return possiblePath.length() >= MIN_REFERENCED_PATH_LENGTH && possiblePath.contains(DOT) && isDescriptorPath(possiblePath);
+    }
+
+    private Optional<String> toAbsolutePath(String currentPath, String relativeOrAbsolutePath) {
+        try {
+            return Optional.of(Paths.get(currentPath).resolve(relativeOrAbsolutePath).normalize().toString());
+        } catch (InvalidPathException e) {
+            // If either path was invalid, ignore and continue.
+            return Optional.empty();
+        }
+    }
+
+    protected EntryType determineType(FileTree fileTree, String path) {
+        Set<EntryType> entryTypes = language.getEntryTypes();
+        // If this descriptor language only supports one type of entry, return it.
+        if (entryTypes.size() == 1) {
+            return entryTypes.iterator().next();
+        }
+        // Otherwise, this method must be overriden.
+        throw new UnsupportedOperationException();
+    }
+
+    protected String determineName(FileTree fileTree, String path) {
+        return null;
+    }
+
+    protected DescriptorLanguageSubclass determineSubclass(FileTree fileTree, String path, EntryType type) {
+        Set<DescriptorLanguageSubclass> subclasses = DescriptorLanguageSubclass.valuesForEntryType(type);
+        // If this descriptor language only supports one type of subclass, return it.
+        if (subclasses.size() == 1) {
+            return subclasses.iterator().next();
+        }
+        // Otherwise, this method must be overriden.
+        throw new UnsupportedOperationException();
+    }
+
+    protected String readFile(FileTree tree, String path) {
+        String content = tree.readFile(path);
+        if (content == null) {
+            throw new CustomWebApplicationException("could not find file", HttpStatus.SC_INTERNAL_SERVER_ERROR);
```

Comment:
Should this be a 500, it isn't immediately clear to me, is there a possibility this just means the workflow referenced a file that legitimately was forgotten to be committed to a repo, a typo, etc. ?

Reply:
Currently, this method is always invoked with a path that we know exists.  So, if we get here, it's probably a bug or server error.  I'm not opposed to changing it to `NOT_FOUND`, though, so I did.  Also added a bit of logging so we can detect this problem in our logstream, if it ever happens.